File: foundation/foundation-lineage/foundation-data-lineage/aissemble-foundation-data-lineage-python/src/aissemble_data_lineage/run_event.py
###
# #%L
# aiSSEMBLE::Foundation::Data Lineage::Python
# %%
# Copyright (C) 2021 Booz Allen
# %%
# This software package is licensed under the Booz Allen Public License. All Rights Reserved.
# #L%
###
from datetime import datetime
from typing import List

from . import Dataset
from .run import Run
from .job import Job
from openlineage.client.run import RunEvent as OpenLineageRunEvent, RunState
from .data_lineage_config import DataLineageConfig

"""
Contains the general concepts needed to perform a base AIOps Reference Architecture Data Action.
A Data Action is a step within a Data Flow.
"""


class RunEvent:
    event_type: str
    event_time: datetime
    run: Run
    job: Job
    inputs: List[Dataset]
    outputs: List[Dataset]
    producer: str
    schema_URL: str

    _config: DataLineageConfig

    def __init__(self, run: Run, job: Job, event_type: str):
        self._config = DataLineageConfig()
        self.schema_URL = self._config.schema_url()
        self.event_time = datetime.utcnow()
        self.producer = self._config.producer(job_name=job.name)
        self.run = run
        self.job = job
        self.inputs = []
        self.outputs = []
        self.event_type = event_type

    def set_inputs(self, inputs: List[Dataset]) -> None:
        """
        :param inputs: List of Datasets to be added to this RunEvent's inputs collection
        """
        self.inputs = inputs

    def get_inputs(self) -> List[Dataset]:
        """
        :return: A copy of this RunEvent's current inputs collection
        """
        return self.inputs.copy()

    def add_input(self, dataset: Dataset) -> None:
        """
        Appends a dataset to the input dataset list
        :param dataset: Dataset to be added to this RunEvent's inputs collection
        """
        self.inputs.append(dataset)

    def set_outputs(self, outputs: List[Dataset]) -> None:
        """
        :param outputs: List of Datasets to be added to this RunEvent's outputs collection
        """
        self.outputs = outputs

    def get_outputs(self) -> List[Dataset]:
        """
        :return: A copy of this RunEvent's current outputs collection
        """
        return self.outputs.copy()

    def add_output(self, dataset: Dataset) -> None:
        """
        Appends a dataset to the output dataset list
        :param dataset: Dataset to be added to this RunEvent's outputs collection
        """
        self.outputs.append(dataset)

    def get_open_lineage_run_event(self) -> OpenLineageRunEvent:
        """
        Converts this aiSSEMBLE RunEvent to its OpenLineage form.
        :return:
        """

        event = OpenLineageRunEvent(
            eventType=RunState[self.event_type],
            eventTime=self.event_time.isoformat(timespec="milliseconds") + "Z",
            run=self.run.get_open_lineage_run(),
            job=self.job.get_open_lineage_job(),
            inputs=[dataset.get_open_lineage_dataset() for dataset in self.inputs],
            outputs=[dataset.get_open_lineage_dataset() for dataset in self.outputs],
            producer=self.producer,
            # TODO: Add schemaURL argument after updating to OpenLineage Client release >0.27.2
            # schemaURL=self.schema_URL
        )
        return event

File: foundation/foundation-lineage/foundation-data-lineage/foundation-data-lineage-java/src/main/java/com/boozallen/aissemble/data/lineage/RunEvent.java
package com.boozallen.aissemble.data.lineage;

/*-
 * #%L
 * aiSSEMBLE::Foundation::Data Lineage Java
 * %%
 * Copyright (C) 2021 Booz Allen
 * %%
 * This software package is licensed under the Booz Allen Public License. All Rights Reserved.
 * #L%
 */

import com.boozallen.aissemble.data.lineage.config.ConfigUtil;
import io.openlineage.client.OpenLineage;
import io.openlineage.client.OpenLineage.RunEvent.EventType;

import java.net.URI;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * The root level object in the structure of a Data Lineage event. In addition to the properties of this event,
 * it also contains all other objects that make up the event's contents.
 */
public class RunEvent {

    private String eventType;
    private Date eventTime;
    private Run run;
    private Job job;
    private String producer;
    private String schemaURL;
    private List<InputDataset> inputs;
    private List<OutputDataset> outputs;
    private static ConfigUtil util = ConfigUtil.getInstance();

    public RunEvent(Run run, Job job, String eventType) {
        this.schemaURL = util.getDataLineageSchemaUrl();
        this.eventTime = new Date();
        this.producer = util.getProducer(job.getName());
        this.run = run;
        this.job = job;
        this.eventType = eventType;
    }

    /**
     * Returns a RunEvent object from the OpenLineage Client library
     * @return A RunEvent
     */
    public OpenLineage.RunEvent getOpenLineageRunEvent() {
        URI producer = URI.create(this.producer);
        OpenLineage openLineage = new OpenLineage(producer);
        ZonedDateTime zTime = this.eventTime.toInstant().atZone(ZoneId.of("UTC"));

        ArrayList<OpenLineage.InputDataset> olInputs = new ArrayList<>();
        if(inputs != null) {
            inputs.forEach(input -> {
                olInputs.add((OpenLineage.InputDataset) input.getOpenLineageDataset());
            });
        }

        ArrayList<OpenLineage.OutputDataset> olOutputs = new ArrayList<>();
        if(outputs != null) {
            outputs.forEach(output -> {
                olOutputs.add((OpenLineage.OutputDataset) output.getOpenLineageDataset());
            });
        }

        return openLineage.newRunEvent(EventType.valueOf(this.eventType),
                zTime,
                run.getOpenLineageRun(),
                job.getOpenLineageJob(),
                !olInputs.isEmpty() ? olInputs : null,
                !olOutputs.isEmpty() ? olOutputs : null);
    }

    /**
     * Accessor for the event type field
     * @return This RunEvent's type
     */
    public String getEventType() {
        return eventType;
    }

    /**
     * Sets the event type field value
     */
    public void setEventType(String eventType) {
        this.eventType = eventType;
    }

    /**
     * Accessor for the event time field
     * @return This RunEvent's type
     */
    public Date getEventTime() {
        return eventTime;
    }

    /**
     * Sets the event time value
     */
    public void setEventTime(Date eventTime) {
        this.eventTime = eventTime;
    }

    /**
     * Accessor for the contained Run object
     * @return The Run contained in this RunEvent
     */
    public Run getRun() {
        return run;
    }

    /**
     * Sets the contained Run to a given Run object
     * @param run The new Run for this RunEvent
     */
    public void setRun(Run run) {
        this.run = run;
    }

    /**
     * Accessor for the contained Job object
     * @return The Job contained in this RunEvent
     */
    public Job getJob() {
        return job;
    }

    /**
     * Sets the contained Job to a given Job object
     * @param job The new Job for this RunEvent
     */
    public void setJob(Job job) {
        this.job = job;
    }

    /**
     * Accessor for the producer field
     * @return This RunEvent's producer
     */
    public String getProducer() {
        return producer;
    }

    /**
     * Sets the producer value
     */
    public void setProducer(String producer) {
        this.producer = producer;
    }

    /**
     * Accessor for the schema used to define this RunEvent's structure
     * @return This RunEvent's schema
     */
    public String getSchemaURL() {
        return schemaURL;
    }

    /**
     * Sets the schema value
     */
    public void setSchemaURL(String schemaURL) {
        this.schemaURL = schemaURL;
    }

    /**
     * Accessor for the inputs Datasets
     * @return The inputs list
     */
    public List<InputDataset> getInputs() {
        return inputs;
    }

    /**
     * Sets the inputs datasets to a given List of values
     * @param inputs The List of inputs for this RunEvent
     */
    public void setInputs(List<InputDataset> inputs) {
        this.inputs = inputs;
    }

    /**
     * Accessor for the outputs Datasets
     * @return The outputs list
     */
    public List<OutputDataset> getOutputs() {
        return outputs;
    }

    /**
     * Sets the outputs datasets to a given List of values
     * @param outputs The List of outputs for this RunEvent
     */
    public void setOutputs(List<OutputDataset> outputs) {
        this.outputs = outputs;
    }

}
